### ISSUE TITLE
Implement stop_on_error in event and handler

### ISSUE THREAD

**Issue by spaceone** [circuits/circuits#128]
Event.stop_on_error and handler.stop_on_error is a very useful thing and should be implemented.

**Comment by prologic** [comment]
Hmm an interesting idea...

Don't like the name per se (*yet*) but let's nut this out...

So what *I think* you're saying is:

- Set ``Event.on_error = "abort"``
- If ``Event.on_error`` is "abort"; don't process other event handlers.

What *should* we do about the error thougH? hide it, log it, still fire the ``exception()`` event? Should we support other kinds of ``on_error`` behavior?

**Comment by spaceone** [comment]
yes, let's still fire the exception() event! I can't image other kinds of on_error behavior but maybe we will have ideas in the future!

**Comment by prologic** [comment]
Perhaps while we're at it; should we consolidate the ``.success``, ``.complete`` and ``.failure`` attributes too? (*NB:I There's a bug with the complete behavior*).

Meaning we break API (*for circuits 3.3*):

- ``on_error``
- ``on_success``
- ``on_complete``

**Comment by spaceone** [comment]
Can you give me an example what value "on_success" could have?
(I don't like to work with strings as it is error prone).

**Comment by prologic** [comment]
Good question; perhaps:

```#!python
e = Event()
e.on_success = "event_to_be_fired_on_auccess", "optional_channel"
```

**Comment by spaceone** [comment]
How do you think about something more encoupled and OOP like the following:
```
class EventOnNothing(object):
»   def task(self, manager, state):
»   »   pass

class OnSuccess(EventOnNothing):
»   def __init__(self, event_name, *channels):
»   »   self.event_name = event_name
»   »   self.channels = channels

»   def task(self, manager, state):
»   »   if self.event_name:
»   »   »   manager.fire(self.event_name, *self.channels)


class OnError(EventOnNothing):
»   def __init__(self, abort=False):
»   »   self.abort = abort

»   def task(self, manager, state):
»   »   if self.abort:
»   »   »   raise StopIteration


class Event(object):  # the circuits.Event class
»   def __init__(self):
»   »   self.on_error = EventOnNothing()
»   »   self.on_success = EventOnNothing()
»   »   self.on_complete = EventOnNothing()
```

**Comment by spaceone** [comment]
Maybe we could outsource behavior in Manager.processTask into some kind of plugins or something. (The method is marked as too complex).

**Comment by prologic** [comment]
Hmm I think what you've presented adds a lot of complexity too as well as overhead for the developer :) I'm not really a big fan of OOP in general; that's why "components" are my preferred pattern! I agree that we should refactor out some of this stuff; specifically I think coroutines should be factored out somehow as well.

**Comment by prologic** [comment]
How about #220 ?

**Comment by prologic** [comment]
REmoving myself as asignee, I'm not working on this sorry 😆
